Mathlib/Topology/MetricSpace/Congruence.lean

```diff
@@ -0,0 +1,219 @@
+/-
+Copyright (c) 2024 Jovan Gerbscheid. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jovan Gerbscheid, Newell Jensen
+-/
+import Mathlib.Topology.MetricSpace.Isometry
+
+/-!
+# Congruences
+
+This file defines `Congruent`, i.e., the equivalence between sets of points in a metric
+space where all corresponding pairwise distances are the same. The motivating example are
+triangles in the plane.
+
+## Main results
+
+In the case of an `EMetricSpace` we show an `IsometryEquiv` between the points:
+`Set.range v₁ ≃ᵢ Set.range v₂`.
+
+## Implementation notes
+
+After considering two possible approaches to defining congruence — either based on equal pairwise
+distances or the existence of an isometric equivalence — we have opted for the broader concept of
+equal pairwise distances. This notion is commonly employed in the literature across various metric
+spaces that lack an isometric equivalence.
+
+For more details see the [Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Euclidean.20Geometry).
+
+## Notation
+
+* `v₁ ≅ v₂`: for `Congruent v₁ v₂`.
+-/
+
+variable {ι ι' : Type*} {P₁ P₂ P₃ : Type*} {v₁ : ι → P₁} {v₂ : ι → P₂} {v₃ : ι → P₃}
+
+noncomputable section
+
+/-- A congruence between indexed sets of vertices v₁ and v₂.
+Use `open scoped Congruence` to access the `v₁ ≅ v₂` notation. -/
+def Congruent (v₁ : ι → P₁) (v₂ : ι → P₂) [PseudoEMetricSpace P₁]
+    [PseudoEMetricSpace P₂] : Prop :=
+  ∀ (i₁ i₂ : ι), (edist (v₁ i₁) (v₁ i₂) = edist (v₂ i₁) (v₂ i₂))
+
+@[inherit_doc]
+scoped[Congruence] infixl:25 " ≅ " => Congruent
+
+/-- A congruence holds if and only if all extended distances are the same. -/
+lemma congruent_iff_edist_eq [PseudoEMetricSpace P₁] [PseudoEMetricSpace P₂] :
+    Congruent v₁ v₂ ↔ (∀ (i₁ i₂ : ι), (edist (v₁ i₁) (v₁ i₂) = edist (v₂ i₁) (v₂ i₂))) :=
+  forall₂_congr (fun _ _ => by simp only [edist])
+
+/-- A congruence holds if and only if all non-negative distances are the same. -/
+lemma congruent_iff_nndist_eq [PseudoMetricSpace P₁] [PseudoMetricSpace P₂] :
+    Congruent v₁ v₂ ↔ (∀ (i₁ i₂ : ι), (nndist (v₁ i₁) (v₁ i₂) = nndist (v₂ i₁) (v₂ i₂))) :=
+  forall₂_congr (fun _ _ => by rw [edist_nndist, edist_nndist]; norm_cast )
```

Comment:
Are you aware of the `peel` tactic? I’m not saying you must use it instead of `forall₂_congr` but in more complicated cases it is a lot easier than wrestling those lemmas.